simple-app.org

```diff
@@ -0,0 +1,223 @@
+* Simple App
+  
+This is an example of a simple Clojure app written using Literate programming
+paradigm using org-mode.
+
+** Setting up the project
+   
+First we setup the project. In the source block bellow you can see that the
+language is set to ~zsh~. This tells babel that this snippet should be executed
+within a shell. I also set the ~:results~ header argument to silent because in
+this case we don't care about the output of running this snippet. 
+
+#+BEGIN_SRC zsh :results silent
+mkdir -p simple-app/src/simple_app
+mkdir -p simple-app/test/simple_app
+mkdir -p simple-app/dev
+#+END_SRC
+
+Now we create our basic deps file. Note that our source block name is set to
+clojure. We do this because that will make the file sintax hilighting work.
+
+To save a snippet like this on a path we set the ~:tangle~ header argument and
+when we run the emacs command ~org-babel-tangle~ our file will be saved to the
+path specified.
+
+#+begin_src clojure :tangle simple-app/deps.edn
+{:deps
+ {org.clojure/clojure {:mvn/version "1.10.1"}}
+ :aliases
+ {:test
+  {:extra-paths ["test"]
+   :extra-deps {com.cognitect/test-runner
+                {:git/url "https://github.com/cognitect-labs/test-runner.git"
+                 :sha "209b64504cb3bd3b99ecfec7937b358a879f55c1"}}
+   :main-opts ["-m" "cognitect.test-runner"]}}}
+#+end_src
+
+** Running and evaluating code
+
+Now that we have our basic project setup we can work on our project namespaces.
+
+First we start by starting a REPL. There are a few different ways you can
+approach this. If you are in emacs you could simply run ~cider-jack-in~ but
+if you notice this file isn't in the same directory as the project. That means
+cider wont know what ~deps.edn~ file to load.  
+
+A solution for this is to manually run your REPL command. Notice the ~:dir~
+header variable bellow. It tells Emacs the relative path of where to execute the
+command. A quirk of doing this is that the following command is blocking and Emacs
+seems to freeze on waiting for output. You can ~C-g~ to unfreeze the editor.
+Notice we set ~:session~ so that our future zsh commands dont get blocked.
+
+#+BEGIN_SRC zsh :dir simple-app :results silent :session simple-app
+clojure -R:test -Sdeps '{:deps {nrepl {:mvn/version "0.7.0"} cider/cider-nrepl {:mvn/version "0.25.2"}}}' -m nrepl.cmdline --middleware '["cider.nrepl/cider-middleware"]'
+#+END_SRC
+
+Lets fetch that nrepl port.
+
+#+BEGIN_SRC zsh :dir simple-app
+cat .nrepl-port
+#+END_SRC
+
+#+RESULTS:
+: 45979
+
+Now we run ~cider-connect-clj~ and we can start working on our project.
+
+Lets start with a simple test.
+
+#+begin_src clojure :ns simple-app.core-test :tangle simple-app/test/simple_app/core_test.clj :results silent
```

Comment:
Does the `ns` form get added to the output file automatically?

Reply:
yes! if you use `:ns simple-app.core-test` in the header. Its a neat trick to save one line :P